keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+resource instance, then NodeUnprepareResource (see below) must have been called
+successfully. This ensures that network-attached resource are available again
+for other Pods, including those that might get scheduled to other nodes. It
+also signals that it is safe to deallocate and delete the ResourceClaim.
+
+#### Communication between kubelet and resource kubelet plugin
+
+Resource kubelet plugins are discovered through the [kubelet plugin registration
+mechanism](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration). A
+new "ResourcePlugin" type will be used in the Type field of the
+[PluginInfo](https://pkg.go.dev/k8s.io/kubelet/pkg/apis/pluginregistration/v1#PluginInfo)
+response to distinguish the plugin from device and CSI plugins.
+
+Under the advertised Unix Domain socket the kubelet plugin provides the following
+gRPC interface. It was inspired by
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
+with “volume” replaced by “resource” and volume specific parts removed.
+
+##### `NodePrepareResource`
+
+This RPC is called by kubelet when a Pod that wants to use the
+specified resource is scheduled on a node.  The Plugin SHALL assume
+that this RPC will be executed on the node where the resource will be
+used.  The Plugin SHALL return device name and kind for allocated
+device[s].
+
+The Plugin SHALL ensure that there are json file[s] in CDI format
+for the allocated resource. These files SHALL be used by runtime to
+update runtime configuration before creating containers that use the
+resource.
+
+This operation SHALL do as little work as possible as it’s called
+after a pod is scheduled to a node. All potentially failing operations
+SHALL be done during allocation phase.
+
+This operation MUST be idempotent. If the resource corresponding to
+the `resource_id` has already been prepared, the Plugin MUST reply `0
+OK`.
+
+If this RPC failed, or kubelet does not know if it failed or not, it
+MAY choose to call `NodePrepareResource` again, or choose to call
+`NodeUnprepareResource`.
+
+On a successful call this RPC should return set of fully qualified
+CDI device names, which kubelet MUST pass to the runtime through the CRI
+protocol.
+
+```protobuf
+message NodePrepareResourceRequest {
+  // The UID of the ResourceClaim. This field is REQUIRED.
+  string resource_uid = 1;
+}
+
+message NodePrepareResourceResponse {
+  // These are the additional devices that kubelet must
+  // make available via the container runtime. A resource
+  // may have zero or more devices.
+  repeated string cdi_device = 1;
+}
+```
+
+CRI protocol MUST be extended for this purpose, e.g. list of CDI
+device ids should be added to the CRI Device structure:
+
+```protobuf
+// Device specifies a host device to mount into a container.
+message Device {
+    ...
+    string permissions = 3;
+    // Set of fully qualified CDI device names
+    repeated string cdi_device = 4;
```

Comment:
@mrunalp please review.

Reply:
I see a distinct lack of sig-node input on this in the last months - is there any way to get some attention for it?

Reply:
@mrunalp Here is an example of cdi_device: intel.com/fpga=Arria10, where intel.com/fpga is a kind and Arria10 is a device name (see more details below). I'll update the KEP shortly.

Here is how fully qualified CDI device name is explained in the [CDI specification](https://github.com/container-orchestrated-devices/container-device-interface/blob/master/SPEC.md):

```
{
    "cdiVersion": "0.3.0",
    "kind": "<name>",

    "devices": [
        {
            "name": "<name>",
...

- kind (string, REQUIRED) field specifies a label which uniquely identifies the device vendor. It can be used to disambiguate the vendor that matches a device, e.g: docker/podman run --device vendor.com/device=foo ....
The kind label has two segments: a prefix and a name, separated by a slash (/).
The name segment is required and must be 63 characters or less, beginning and ending with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between.
The prefix must be a DNS subdomain: a series of DNS labels separated by dots (.), not longer than 253 characters in total, followed by a slash (/).
Examples (not an exhaustive list):
Valid: vendor.com/foo, foo.bar.baz/foo-bar123.B_az.
Invalid: foo, vendor.com/foo/, vendor.com/foo/bar.

CDI Devices
The devices field describes the set of hardware devices that can be requested by the container runtime user. Note: For a CDI file to be valid, at least one entry must be specified in this array.

- devices (array of objects, REQUIRED) list of devices provided by the vendor.
name (string, REQUIRED), name of the device, can be used to refer to it when requesting a device.
Beginning and ending with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between.
e.g: docker/podman run --device foo ...
Entries in the array MUST use the same schema as the entry for the name field

```

Reply:
A PR adding some more clarifications to the KEP text is in https://github.com/pohly/enhancements/pull/15/files

@mrunalp does that resolve your question? Then I'll take that change into my branch and mark this conversation as resolved.